orc8r/cloud/go/tools/migrations/m010_default_apns/main.go

```diff
@@ -0,0 +1,452 @@
+
+	err = mapSubscribersToAPN(tx, builder, apn.pk, subPKs)
+	if err != nil {
+		return err
+	}
+
+	err = mergeGraphs(tx, builder, network, allGIDs)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getSubscribersMissingAPN(tx *sql.Tx, builder sqorc.StatementBuilder, network string) ([]ent, error) {
+	// SELECT a.graph_id,a.pk, b.pk,b.type
+	// FROM cfg_entities AS a
+	// LEFT JOIN cfg_assocs ON a.pk=from_pk
+	// LEFT JOIN cfg_entities AS b ON to_pk=b.pk
+	// WHERE a.network=network AND a.type='subscriber';
+	tblA, tblB := "a", "b"
+	nidColA := makeCol(tblA, entNidCol)
+	gidColA := makeCol(tblA, entGidCol)
+	pkColA, pkColB := makeCol(tblA, entPkCol), makeCol(tblB, entPkCol)
+	typeColA, typeColB := makeCol(tblA, entTypeCol), makeCol(tblB, entTypeCol)
+
+	rows, err := builder.
+		Select(gidColA, pkColA, pkColB, typeColB).
+		From(fmt.Sprintf("%s AS %s", entityTable, tblA)).
+		LeftJoin(fmt.Sprintf("%s ON %s=%s", entityAssocTable, pkColA, aFrCol)).
+		LeftJoin(fmt.Sprintf("%s AS %s ON %s=%s", entityTable, tblB, aToCol, pkColB)).
+		Where(
+			squirrel.And{
+				squirrel.Eq{nidColA: network},
+				squirrel.Eq{typeColA: subscriberEntType},
+			},
+		).
+		RunWith(tx).Query()
+	if err != nil {
+		return nil, errors.Wrap(err, "getSubscribersMissingAPN: select subscribers")
+	}
+
+	assocs := map[ent][]ent{}
+	for rows.Next() {
+		a := ent{typ: subscriberEntType}
+		pkB, typeB := &sql.NullString{}, &sql.NullString{}
+		err = rows.Scan(&a.graphID, &a.pk, pkB, typeB)
+		if err != nil {
+			return nil, errors.Wrap(err, "getSubscribersMissingAPN: scan subscriber")
+		}
+		assocs[a] = append(assocs[a], ent{pk: pkB.String, typ: typeB.String})
+	}
+	err = rows.Err()
+	if err != nil {
+		return nil, errors.Wrap(err, "getSubscribersMissingAPN: SQL rows error")
+	}
+
+	var subsMissingAPN []ent
+	for sub, ents := range assocs {
+		hasAPN := false
+		for _, ent := range ents {
+			if ent.typ == apnEntType {
+				hasAPN = true
+			}
+		}
+		if !hasAPN {
+			subsMissingAPN = append(subsMissingAPN, sub)
+		}
+	}
+
+	return subsMissingAPN, nil
+}
+
+func mapSubscribersToAPN(tx *sql.Tx, builder sqorc.StatementBuilder, apnPK string, subscriberPKs []string) error {
+	b := builder.Insert(entityAssocTable).Columns(aFrCol, aToCol)
+	for _, subPK := range subscriberPKs {
+		b = b.Values(subPK, apnPK)
+	}
+	sqlStr, args, _ := b.ToSql()
+	glog.Infof("[RUN] %s %v", sqlStr, args)
+	_, err := b.RunWith(tx).Exec()
+	if err != nil {
+		return errors.Wrap(err, "mapSubscribersToAPN: insert error")
+	}
+
+	return nil
+}
+
+// mergeGraphs merges the graphs for all passed graph IDs.
+// Assumes gids is non-empty.
+func mergeGraphs(tx *sql.Tx, builder sqorc.StatementBuilder, network string, gids []string) error {
+	gids = funk.UniqString(gids)
+	sort.Strings(gids)
+	mergedGID, oldGIDs := gids[0], gids[1:]
+
+	isOldGID := squirrel.Or{}
+	for _, gid := range oldGIDs {
+		isOldGID = append(isOldGID, squirrel.Eq{entGidCol: gid})
+	}
```

Comment:
you should be able to do `squirrel.Eq{entGidCol: oldGIDs}` which will render to a `WHERE... IN` clause